### ISSUE TITLE
✨ Add a cancel button on the plugin install dialog

### ISSUE THREAD

**Issue by goanpeca** [napari/napari#3070]
## Motivation

Currently when a package is being installed and the dialog is closed (via close button or via close icon on the window corenr tools) the install process will be terminated without any notification.

## Pitch

When a user closes the plugin install dialog and something is being installed, I would like to have some clearer notification in order to effectively continue or close.

## Alternatives

To provide a better UI/UX we could whenever a process is still running on close we could:
- Prompt a message box on close whenever there is a process still running (yes/n)
- change the button from close to `cancel` 

I prefer the dialog since that makes it more clear, even when closing with the `X` on the corner window toolbars, so I would favor that approach

## Additional context

This behavior could change, if we move the plugin dialog from being a dialog to being a sidebar/dockwidget in which case the concept of "closing" will no longer exist.

There might be additional work to do to communicate the state of installation via the notification manager

---

I would really like to hear opinions from @napari/core-devs, @liaprins and others

**Comment by liaprins-czi** [comment]
Great idea for an improvement!

> I prefer the dialog since that makes it more clear, even when closing with the X on the corner window toolbars, so I would favor that approach

Yes I agree with this. But I also like the idea of changing the close button to be called "Cancel", because users may actually _want_ to cancel sometimes, but not now how or know that it is an option to do by closing the dialog. I think that doing both changes you suggest would be beneficial.

**Comment by goanpeca** [comment]
Thanks for the feedback @liaprins-czi 🤗

**Comment by liaprins-czi** [comment]
### Overview of design need
- **In a sentence or two, describe the problem to be solved for users**
When a user closes the plugin install dialog and a plugin is being installed, they should be more clearly made aware of the implications as well as alternatives.

### What level of design is needed? (Choose all that apply)
- [ ] **User experience research:**  high-level recommendation/exploration of user needs, design heuristics, and / or best practices to inform a design experience  
(Use this option when you feel there’s a challenge to be solved, but you’re curious about what the experience should be — may involve research studies to understand challenges/opportunities for design)
- [X] **information flow / conceptual:** organizing and structuring of information flow and content, including layout on screen or across multiple steps
- [X] **Visual:** creating mockups, icons, etc
(If choosing this level alone, it means that the content to be mocked up and its organization is already known and specified)

### Is design a blocker?
- [ ] **Yes:** engineering cannot proceed without a design first
- [X] **No:** engineering can create a first version, and design can come in later to iterate and refine

If selecting **Yes**, how much design input is needed to unblock engineering? For example, is a full, final visual design needed, or just a recommendation of which conceptual direction to go?

### T-shirt size
- [ ] ●○○○○ XS: able to be done in a typical workday or two
- [X] ●●○○○ S: able to do within one sprint, if also working on other things in the sprint
- [ ] ●●●○○ M: able to do within one sprint, if it is the sole focus
- [ ] ●●●●○ L: will likely take more than one sprint to complete, but less than two
- [ ] ●●●●● XL: a very large experience to be determined or overhauled, will need to be broken down further

**Comment by isabela-pf** [comment]
Hi @goanpeca! I'm planning to work on the visuals for this. 

To clarify, are you talking about closing this plugins dialog while installing a plugin? 
<img width="1192" alt="Screen Shot 2021-08-31 at 2 09 31 PM" src="https://user-images.githubusercontent.com/50221806/131576380-a1f84093-a8af-40f1-8e63-b6f50d155768.png">

Since I'm new here, all of these sound like reasonable options. I haven't found too many dialogs in napari yet. Are they typically used for menus like this? It seems like more things appear in the left side window area.

**Comment by sofroniewn** [comment]
> Since I'm new here, all of these sound like reasonable options. I haven't found too many dialogs in napari yet. Are they typically used for menus like this? It seems like more things appear in the left side window area.

@jni do you have any preferences here? also might be good for @jni and @isabela-pf to meet :-)

**Comment by goanpeca** [comment]
Hi @isabela-pf :)

So a bit of context:

### The present

When Plugin dialog  is opened you can now install and uninstall several things, but if you hit the close button, or the `x` on the dialog title bar it will just stop the process with no warning.

### Options

1.) We can add a popup so that if a user tries to close the dialog while stuff is going on (install, update, uninstall) it will warn the user and maybe provide a yes/no to confirm closing it which will continue to stop whatever action was taking place. This is a simpler faster solution that would provide feedback but would yield a dialog on to of a dialog which I believe is not great from a UX/UI perspective, but gets the job done :) 

2.) Slightly more involved, but we could not kill the processes, but just hide the dialog and make something appear either on the status bar, or the notification manager so that a user knows something is going on, or even inform when it finished with the notification manager

### Possible future

The plugin dialog manager would become a dock on the side, so the concept of a dialog would disappear and there would not be any `close` button anymore. This would also preclude the use of the notification manager since the plugin "panel" (let's call it like that for now) would be there, in a similar way as VSCode. Or we could also use the notification manager to informe something finished installing if the plugin panel is not visible, or always display that message no matter what. 

<img width="424" alt="Screen Shot 2021-08-31 at 5 02 25 PM" src="https://user-images.githubusercontent.com/3627835/131581555-b4381056-1254-4a76-b261-e2571f1734e6.png">

**Comment by isabela-pf** [comment]
Thanks everyone for the insight! I don’t know how difficult it may be to make some of these things happen, so I’ve provided a few options. Here’s my thoughts so far:

## Option 1

My ideal is to not have closing the dialog stop all installations. 

I’d also like to see more clearly which plugins are being installed or queued for install, so I added them to the `Installed Plugin` section with a label that they are being installed. This also adds the requested `cancel` button. 

<img width="1081" alt="napari's plugin manager with a new plugin being installed in the Installed Plugin section" src="https://user-images.githubusercontent.com/50221806/132072125-b7caed73-92cc-4471-b0ad-d2df1f320d97.png">

It’s a tiny interaction, but you can [try it in Figma](https://www.figma.com/file/5fZxK6UnbOGSHrgELY1a03/Plugin-Installation?node-id=7%3A102) if you’d like.

While we don’t have any other kind of progress bar system that I’m aware of, I think it’s feasible to add an item in the Plugins menu for people to check without having to open the whole interface again. I’d prefer this be a temporary option, though.

<img width="532" alt="The Plugins menu with Installing 1 of 1 added" src="https://user-images.githubusercontent.com/50221806/132072162-8c36bac1-5188-459b-8e40-0f2111c50a64.png">

## Option 2
Follow the same pattern as Option 1, but make it so that the dialog can’t be closed while a plugin is being installed or while there are queued installs. 

I’m thinking of disabling the close button until installation is complete. Macs make it look like this:
<img width="170" alt="A disabled close button in the napari Plugins manager" src="https://user-images.githubusercontent.com/50221806/132072314-2bf69d87-766a-4139-97fe-8f9b6de50df5.png">

## Option 3

Have the double dialog approach mentioned above. This means when a user closes the Plugins manager while a plugin is being installed, they are interrupted by another dialog clarifying that this will stop all current and queued installations.

This would be my least favorite option by far and I would only want it as a temporary option if implementing a better experience is going to take longer than we expect. Stacking dialogs on top of each other isn’t a good experience, but some feedback is probably better than none in this case.

## Additional notes
- Once we end up with some kind of progress bar system like @liaprins-czi has suggested at #3230, I think it may make sense for plugin installation to trigger a progress bar for consistency’s sake.
- I didn’t include any options in napari’s left side bar because it seemed like the wrong grouping of information. It looks to me like the side bar is used to manage and manipulate the open image while dialogs like this and Preferences are used to manage the whole of napari. Please let me know if I’m misunderstanding.

**Comment by liaprins-czi** [comment]
Love all of these! 

> Once we end up with some kind of progress bar system like @liaprins-czi has suggested at ✨ Designs for progress bars #3230, I think it may make sense for plugin installation to trigger a progress bar for consistency’s sake.

Yes I think this is agreat idea!

> I didn’t include any options in napari’s left side bar because it seemed like the wrong grouping of information. It looks to me like the side bar is used to manage and manipulate the open image while dialogs like this and Preferences are used to manage the whole of napari. Please let me know if I’m misunderstanding.

Yes, completely correct!

**Comment by goanpeca** [comment]
Hi @isabela-pf thanks for the feedback

I think we could implement option 1 and option 2

I do not think that a plugin that is being installed should move to the installed plugins sections, cause it is effectively not installed, 
<img width="782" alt="Screen Shot 2021-09-08 at 6 23 56 PM" src="https://user-images.githubusercontent.com/3627835/132598303-1e1f3d31-bbf1-49f8-8ab2-33d72caf13d4.png">

I would leave it in the bottom menu, and when it finished it will move to the top menu.

Also one might argue that since things are installed in a queue, then if I clicked on 5 things and want to cancel everything, having to go through each selected plugin and cancel, feels odd. As a user I would expect to have an option to cancel everything with a single option. Option 3, even if the least favorite, does provide that

**Comment by isabela-pf** [comment]
> I think we could implement option 1 and option 2

So it isn't currently possible to have the closing the dialog not cancel installation? I think option 2 will work fine, just want to make sure I understand why choices are being made.

> I do not think that a plugin that is being installed should move to the installed plugins sections, cause it is effectively not installed,

That makes sense. My logic was that it's pretty likely that users will have less extensions installed (and thus less listed) than all the ones available for installation, so it's easier to find all pending installations in the upper box than the lower one. Is there another way to make pending installations easier to find in the existing areas? 

Would it be possible to have all pending installs move to the top of the uninstalled list and be grouped together some way once they've been selected? Without the progress bars, I think it could be hard to see all installs queued if they stay in the long alphabetical order list.

**Comment by goanpeca** [comment]
>So it isn't currently possible to have the closing the dialog not cancel installation?

In case something goes wrong (maybe a really slow internet) I find the option of not being to close the dialog a bit unfriendly to the user. So option 2 feels like an option that could leave a user with an unclosable dialog.

>That makes sense. My logic was that it's pretty likely that users will have less extensions installed (and thus less listed) than all the ones available for installation, so it's easier to find all pending installations in the upper box than the lower one. Is there another way to make pending installations easier to find in the existing areas?

>Would it be possible to have all pending installs move to the top of the uninstalled list and be grouped together some way once they've been selected? Without the progress bars, I think it could be hard to see all installs queued if they stay in the long alphabetical order list.

I could move the pending ones to the top of the bottom list, I feel that could group them better but still leave them at the bottom list. That was going to be my suggestion.

---

Still I feel like we need also option 3, even if it is the least favorite cause it can offer 2 things:
- A way out for whatever reason
- A way to cancel ALL actions happening 

There should be a way out for users, so after all I am not so convinced now with option 2.

**Comment by isabela-pf** [comment]
> I could move the pending ones to the top of the bottom list, I feel that could group them better but still leave them at the bottom list. That was going to be my suggestion.

Okay!

> - A way out for whatever reason
> - A way to cancel ALL actions happening

I agree! I also want a way out, so we just need to make sure that users can close the dialog without cancelling installations in the same interaction, right? (unless I’m misunderstanding)

I'm fine without option 2, but I don't want option 3 as a permanent solution. I'm skeptical about stacking more windows even though the install dialog and a warning dialog are kind of two different things, so what about having a way to cancel without the dialogs?

 Here's a new set of mockups with the installing extension and a cancel all button (in one of two places). 

## Cancel button near search

<img width="1081" alt="Add installation state and cancel button 3" src="https://user-images.githubusercontent.com/50221806/132770141-ca4d9f17-0e25-40e6-b3d1-600820066e30.png">

## Cancel button near install queue

<img width="1081" alt="Add installation state and cancel button 4" src="https://user-images.githubusercontent.com/50221806/132770154-2a38d68d-4303-4931-b0cd-8f0225938887.png">


I’m not certain about the `cancel all installations` wording since I think this would cancel installs and updates, but that seemed wordier than necessary. I wasn’t sure that `processes` was clear enough either. Thoughts?

**Comment by goanpeca** [comment]
Thanks for the updates @isabela-pf, I already have something in place that implements your suggestions :)
